src/toil/jobStores/aws/jobStore.py

```diff
@@ -410,126 +421,90 @@ def getEmptyFileStoreID(self, jobStoreID=None, cleanup=False, basename=None):
-                log.debug("Can not use multipart uploading for %s, uploading whole file at once", dstKey.name)
-                dstKey.set_contents_from_string(readable.read())
-        finally:
-            dstKey.bucket.connection.close()
+        dstObj = cls._getObjectForUrl(url)
+        resource = get_boto3_session().resource('s3')
+
+        log.debug("Uploading %s", dstObj.key)
+        # uploadFile takes care of using multipart upload if the file is larger than partSize (default to 5MB)
+        uploadFile(readable=readable, resource=resource, bucketName=dstObj.bucket_name,
+                   fileID=dstObj.key, partSize=5 * 1000 * 1000)
 
     @staticmethod
-    def _getKeyForUrl(url, existing=None):
+    def _getObjectForUrl(url, existing=None):
         """
-        Extracts a key from a given s3:// URL. On return, but not on exceptions, this method
-        leaks an S3Connection object. The caller is responsible to close that by calling
-        key.bucket.connection.close().
+        Extracts a key (object) from a given s3:// URL.
 
         :param bool existing: If True, key is expected to exist. If False, key is expected not to
-               exists and it will be created. If None, the key will be created if it doesn't exist.
+        exists and it will be created. If None, the key will be created if it doesn't exist.
 
-        :rtype: Key
+        :rtype: S3.Object
         """
         keyName = url.path[1:]
         bucketName = url.netloc
 
         # Get the bucket's region to avoid a redirect per request
-        try:
-            with closing(boto.connect_s3()) as s3:
-                location = s3.get_bucket(bucketName).get_location()
-                region = bucket_location_to_region(location)
-        except S3ResponseError as e:
-            if e.error_code == 'AccessDenied':
-                s3 = boto.connect_s3()
-            else:
-                raise
-        else:
-            # Note that caller is responsible for closing the connection
-            s3 = boto.s3.connect_to_region(region)
+        region = AWSJobStore.__getBucketRegion(bucketName)
 
+        s3 = get_boto3_session().resource('s3', region_name=region)
+        obj = s3.Object(bucketName, keyName)
+        objExists = True
         try:
-            bucket = s3.get_bucket(bucketName)
-            key = bucket.get_key(keyName.encode('utf-8'))
-            if existing is True:
-                if key is None:
-                    raise RuntimeError("Key '%s' does not exist in bucket '%s'." %
-                                       (keyName, bucketName))
-            elif existing is False:
-                if key is not None:
-                    raise RuntimeError("Key '%s' exists in bucket '%s'." %
-                                       (keyName, bucketName))
-            elif existing is None:
-                pass
+            obj.load()
+        except ClientError as e:
+            if e.response.get('Error', {}).get('Code') == '404':
```

Comment:
Are you sure that this shouldn't be:

```
if e.response.get('ResponseMetadata', {}).get('HTTPStatusCode') == '404':
```

?

I would expect a human readable error from `e.response.get('Error', {}).get('Code')`, like `BucketNotFound`.

See: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html